### ISSUE TITLE
`hafnian` fails to backward if construction of derivative graph is required

### ISSUE THREAD

**Issue by Jooyuza** [TuringQ/deepquantum#133]
- If  `torch.autograd.grad` parameter `create_graph` is set as `True`, backward of loss function(output of GBS) fails. This results in failure of derivative graph based features like `hessian`.

- Through examination, this issue relates to the `torch.linalg.eigvals` of hafnian calculation. However, `torch.linalg.eigvals` only will not cause this error, the root cause requires further investigation.

- Minimal Reproduction

1. hessian
```
from torch.autograd.functional import hessian
def f(params):
    cir = dq.QumodeCircuit(nmode=2, cutoff=3, init_state='vac', backend='gaussian')
    cir.s(wires=[0], r=0.8, theta=0)
    cir.s(wires=[1], r=0.8, theta=0)

    cir.bs([0,1])
    cir.ps([0], encode=True)
    cir.bs([0,1])

    d = cir(params, is_prob=True)
    sorted_values = [v for k, v in d.items()][:2]
    probs = torch.cat(sorted_values)
    return probs.sum()

x = torch.tensor([1.0], requires_grad=True)
hessian(f, x)
```

<img width="703" height="430" alt="Image" src="https://github.com/user-attachments/assets/b3bf1f0a-0a0f-43ef-98eb-c118bab52cea" />

Random matrix with `torch.linalg.eigvals` only will not cause this error:
<img width="432" height="214" alt="Image" src="https://github.com/user-attachments/assets/0123c9c5-b64a-4c05-9de8-49d090ad4b4e" />

2. backward fails when `create_graph=True`
```
cir = dq.QumodeCircuit(nmode=2, cutoff=3, init_state='vac', backend='gaussian')
cir.s(wires=[0], r=0.8, theta=0)
cir.s(wires=[1], r=0.8, theta=0)

cir.bs([0,1])
cir.ps([0], encode=True)
cir.bs([0,1])

optimizer = optim.Adam(cir.parameters(), lr=0.01)
epochs = 3

for epoch in range(epochs):
    optimizer.zero_grad()

    t = torch.tensor([1.0], requires_grad=True)

    d = cir(t, is_prob = True)
    sorted_values = [v for k, v in d.items()][:2]
    u = torch.cat(sorted_values).sum()

    du_dt = torch.autograd.grad(
        outputs=u, inputs=t,
        create_graph=True, retain_graph=True
    )[0]

    loss = du_dt + u
    # torch.autograd.set_detect_anomaly(True)
    # with torch.autograd.detect_anomaly():
    #     loss.backward()
    loss.backward()
    optimizer.step()

    if epoch % 1 == 0:
        print(f"Epoch {epoch}: Loss {loss.item():.6f}")
```

<img width="1088" height="333" alt="Image" src="https://github.com/user-attachments/assets/5fe9fa51-874f-48d9-9bb3-0a33e1592bd8" />

 In such cases, `create_graph` can only be set as `True`, otherwise the backward of loss will neglect the derivative part. (verified numerically✔)
